File: frontend/src/lib/types/analysis.ts
import type { AnalysisMode } from "@/lib/analysis/analysis-modes";

export type AnalysisRun = {
  analysis_mode: AnalysisMode;
  config: Record<string, unknown>;
  created_at: string;
  id: number;
  status: string;
};

export type InputDocument = {
  analysis_run_id: number;
  file_hash: string;
  file_path: string;
  id: number;
  original_filename: string;
  tipo: string;
};

export type ExtractedField = {
  bbox: Record<string, number> | null;
  document_filename: string;
  document_page_id: number | null;
  document_tipo: string;
  field_name: string;
  id: number;
  input_document_id: number | null;
  normalized_value: string;
  page: number | null;
  raw_value: string;
};

export type PackageSummary = {
  alerts: PackageSummaryAlert[];
  documents: PackageSummaryDocument[];
  identity: {
    bairro: string | null;
    client: string | null;
    date: string | null;
    disciplines: string[];
    project_code: string | null;
    volumes: string[];
    work_name: string | null;
  };
  stats: {
    document_count: number;
    ld_count: number;
    page_count: number;
    volume_count: number;
  };
};

export type PackageSummaryAlert = {
  message: string;
  severity: string;
};

export type PackageSummaryDocument = {
  classification: string;
  detected_project_codes: string[];
  discipline: string | null;
  document_id: number;
  filename: string;
  ld_pages: number[];
  page_count: number;
  tipo: string;
  tomo: string | null;
  volume: string | null;
};

export type AuditSummary = {
  highlights: AuditSummaryHighlight[];
  metrics: AuditSummaryMetrics;
  sources: AuditSummarySource[];
  status: AuditSummaryStatus;
};

export type AnalysisSignoff = {
  analysis_run_id: number;
  comment: string;
  created_at: string;
  final_status_code: string;
  final_status_label: string;
  reviewer_name: string;
  updated_at: string;
};

export type AuditSummaryStatus = {
  code: string;
  label: string;
  summary: string;
  tone: string;
};

export type AuditSummaryHighlight = {
  message: string;
  tone: string;
};

export type AuditSummaryMetrics = {
  active_issue_count: number;
  attention_count: number;
  dismissed_issue_count: number;
  document_count: number;
  incomplete_count: number;
  inconclusive_issue_count: number;
  issue_count: number;
  ld_row_count: number;
  page_count: number;
  pending_review_count: number;
  relevant_count: number;
  resolved_issue_count: number;
  reviewed_issue_count: number;
  sheet_count: number;
  undeclared_sheet_count: number;
};

export type AuditSummarySource = {
  active_count: number;
  attention_count: number;
  dismissed_count: number;
  incomplete_count: number;
  inconclusive_count: number;
  item_count: number;
  label: string;
  pending_review_count: number;
  relevant_count: number;
  resolved_count: number;
  reviewed_count: number;
  source: string;
  summary: string;
  undeclared_sheet_count: number;
};

export type DirectedModeOutput = {
  entries: DirectedModeEntry[];
  expected: string | null;
  field_label: string | null;
  mode: string;
  query: string | null;
  replace: string | null;
  stats: DirectedModeStats;
  summary: string;
  title: string;
};

export type DirectedModeStats = {
  divergent_count: number;
  document_count: number;
  matching_count: number;
  occurrence_count: number;
  page_count: number;
};

export type DirectedModeEntry = {
  bbox: Record<string, number> | null;
  context: string;
  document_id: number | null;
  expected_value: string | null;
  field_name: string | null;
  filename: string;
  kind: string;
  page: number | null;
  replacement_preview: string | null;
  severity: string;
  value: string;
};

export type PackageMap = {
  documents: PackageMapDocument[];
  identity: PackageSummary["identity"];
  stats: {
    document_count: number;
    ld_section_count: number;
    section_count: number;
    sheet_count: number;
  };
};

export type PackageMapDocument = {
  classification: string;
  discipline: string | null;
  document_id: number;
  filename: string;
  page_count: number;
  sections: PackageMapSection[];
  tipo: string;
  tomo: string | null;
  volume: string | null;
};

export type PackageMapSection = {
  document_id: number;
  document_filename: string;
  end_page: number;
  ld_codes: string[];
  ld_page: number | null;
  ld_row_count: number;
  scope_id: number;
  section_label: string;
  section_type: string;
  sheet_codes: string[];
  sheet_count: number;
  start_page: number;
  title: string;
};

export type AiReview = {
  ai_model: string | null;
  ai_narrative: string | null;
  contexts: AiReviewContext[];
  identity: PackageSummary["identity"];
  mode: string;
  provider_status: string;
  summary: string;
  suggestions: AiReviewSuggestion[];
  stats: {
    context_count: number;
    needs_review_count: number;
    probable_issue_count: number;
    suggestion_count: number;
  };
};

export type AiReviewContext = {
  document_id: number;
  evidence_text: string;
  filename: string;
  kind: string;
  page_end: number;
  page_start: number;
  section_label: string;
  title: string;
};

export type AiReviewSuggestion = {
  category: string;
  message: string;
  reason: string;
  severity: string;
  source: string;
};

export type PageMap = {
  documents: PageMapDocument[];
  stats: {
    document_count: number;
    low_confidence_count: number;
    page_count: number;
    page_type_counts: Record<string, number>;
  };
};

export type PageMapDocument = {
  document_id: number;
  filename: string;
  page_count: number;
  pages: PageMapPage[];
  tipo: string;
};

export type PageMapPage = {
  confidence: string;
  discipline_code: string | null;
  discipline_label: string | null;
  discipline_type: string | null;
  document_id: number;
  evidence_text: string;
  filename: string;
  page: number;
  page_type: string;
  page_type_label: string;
  scope_id: number | null;
  signals: string[];
};

export type FooterAudit = {
  findings: FooterAuditFinding[];
  identity: {
    project_code: string | null;
    work_name: string | null;
  };
  occurrences: FooterAuditOccurrence[];
  stats: {
    document_count: number;
    footer_page_count: number;
    needs_review_count: number;
    occurrence_count: number;
    probable_issue_count: number;
  };
};

export type FooterAuditOccurrence = {
  document_id: number;
  field: string;
  field_label: string;
  filename: string;
  normalized_value: string;
  page: number;
  source_text: string;
  value: string;
};

export type FooterAuditFinding = {
  category: string;
  field: string;
  message: string;
  occurrences: FooterAuditOccurrence[];
  reason: string;
  severity: string;
};

export type DrawingLists = {
  alerts: DrawingListAlert[];
  lists: DrawingListDocument[];
  stats: {
    alert_count: number;
    document_count: number;
    row_count: number;
  };
};

export type DrawingListDocument = {
  document_id: number;
  filename: string;
  project_codes: string[];
  row_count: number;
  rows: DrawingListRow[];
  tipo: string;
};

export type DrawingListRow = {
  description: string;
  document_code: string;
  item: string;
  page: number;
  source_text: string;
};

export type DrawingListAlert = {
  description: string;
  document_code: string;
  filename: string;
  item: string;
  message: string;
  page: number;
  severity: string;
  source_text: string;
  type: string;
};

export type DetectedSheets = {
  documents: DetectedSheetDocument[];
  stats: {
    document_count: number;
    sheet_count: number;
  };
};

export type DetectedSheetDocument = {
  document_id: number;
  filename: string;
  sheet_count: number;
  sheets: DetectedSheet[];
  tipo: string;
};

export type DetectedSheet = {
  description: string | null;
  item: string | null;
  page: number;
  sheet_code: string;
  source_text: string;
};

export type LdSheetCrosscheck = {
  reverse_results: DetectedSheetCrosscheckResult[];
  results: LdSheetCrosscheckResult[];
  stats: {
    attention_count: number;
    compatible_count: number;
    combined_extraction_limit_count: number;
    combined_needs_review_count: number;
    combined_probable_issue_count: number;
    extraction_limit_count: number;
    needs_review_count: number;
    ok_count: number;
    probable_issue_count: number;
    relevant_count: number;
    reverse_extraction_limit_count: number;
    reverse_needs_review_count: number;
    reverse_other_document_count: number;
    reverse_other_section_count: number;
    reverse_probable_issue_count: number;
    reverse_total_count: number;
    total_count: number;
    undeclared_sheet_count: number;
  };
};

export type LdSheetCrosscheckResult = {
  category: string;
  ld_description: string;
  ld_document_code: string;
  ld_filename: string;
  ld_item: string;
  ld_page: number;
  ld_scope_id: number | null;
  ld_source_text: string;
  matched_sheet: LdSheetMatchedSheet | null;
  message: string;
  reason: string;
  severity: string;
  type: string;
};

export type LdSheetMatchedSheet = {
  description: string | null;
  filename: string;
  item: string | null;
  page: number;
  scope_id: number | null;
  sheet_code: string;
  source_text: string;
};

export type LdMatchedRow = {
  description: string;
  document_code: string;
  filename: string;
  item: string;
  page: number;
  scope_id: number | null;
  source_text: string;
};

export type DetectedSheetCrosscheckResult = {
  category: string;
  matched_ld_row: LdMatchedRow | null;
  message: string;
  reason: string;
  severity: string;
  sheet_code: string;
  sheet_description: string | null;
  sheet_filename: string;
  sheet_item: string | null;
  sheet_page: number;
  sheet_scope_id: number | null;
  sheet_source_text: string;
  type: string;
};

export type MemorialAudit = {
  findings: MemorialAuditFinding[];
  identity: {
    bairro: string | null;
    municipality: string | null;
    project_code: string | null;
    work_name: string | null;
  };
  occurrences: MemorialAuditOccurrence[];
  stats: {
    document_count: number;
    extraction_limit_count: number;
    needs_review_count: number;
    occurrence_count: number;
    probable_issue_count: number;
  };
};

export type MemorialAuditOccurrence = {
  document_id: number;
  field: string;
  field_label: string;
  filename: string;
  normalized_value: string;
  page: number;
  source_text: string;
  value: string;
};

export type MemorialAuditFinding = {
  category: string;
  field: string;
  message: string;
  occurrences: MemorialAuditOccurrence[];
  reason: string;
  severity: string;
};

File: backend/app/schemas/analysis.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.core.analysis_modes import ANALYSIS_MODE_DEFAULT


class AnalysisCreateSchema(BaseModel):
    analysis_mode: str = ANALYSIS_MODE_DEFAULT
    config: dict[str, Any] = Field(default_factory=dict)


class AnalysisRunSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: str
    analysis_mode: str
    config: dict[str, Any]
    created_at: datetime


class InputDocumentSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    analysis_run_id: int
    tipo: str
    original_filename: str
    file_path: str
    file_hash: str


class DocumentPageSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_id: int
    page_number: int


class TextSpanSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_page_id: int
    text: str
    bbox: Any | None


class ExtractedFieldSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    input_document_id: int | None
    document_page_id: int | None
    field_name: str
    raw_value: str
    normalized_value: str
    bbox: Any | None


class ExtractedFieldWithContextSchema(ExtractedFieldSchema):
    page: int | None
    document_filename: str
    document_tipo: str


class PackageSummaryIdentitySchema(BaseModel):
    project_code: str | None
    work_name: str | None
    bairro: str | None
    client: str | None
    date: str | None
    volumes: list[str]
    disciplines: list[str]


class PackageSummaryDocumentSchema(BaseModel):
    classification: str
    detected_project_codes: list[str]
    discipline: str | None
    document_id: int
    filename: str
    ld_pages: list[int]
    page_count: int
    tipo: str
    tomo: str | None
    volume: str | None


class PackageSummaryStatsSchema(BaseModel):
    document_count: int
    ld_count: int
    page_count: int
    volume_count: int


class PackageSummaryAlertSchema(BaseModel):
    severity: str
    message: str


class PackageSummarySchema(BaseModel):
    identity: PackageSummaryIdentitySchema
    documents: list[PackageSummaryDocumentSchema]
    stats: PackageSummaryStatsSchema
    alerts: list[PackageSummaryAlertSchema]


class AuditSummaryStatusSchema(BaseModel):
    code: str
    label: str
    summary: str
    tone: str


class AuditSummaryHighlightSchema(BaseModel):
    message: str
    tone: str


class AuditSummaryMetricsSchema(BaseModel):
    active_issue_count: int
    attention_count: int
    dismissed_issue_count: int
    document_count: int
    incomplete_count: int
    inconclusive_issue_count: int
    issue_count: int
    ld_row_count: int
    page_count: int
    pending_review_count: int
    relevant_count: int
    resolved_issue_count: int
    reviewed_issue_count: int
    sheet_count: int
    undeclared_sheet_count: int


class AuditSummarySourceSchema(BaseModel):
    active_count: int
    attention_count: int
    dismissed_count: int
    incomplete_count: int
    inconclusive_count: int
    item_count: int
    label: str
    pending_review_count: int
    relevant_count: int
    resolved_count: int
    reviewed_count: int
    source: str
    summary: str
    undeclared_sheet_count: int = 0


class AuditSummarySchema(BaseModel):
    highlights: list[AuditSummaryHighlightSchema]
    metrics: AuditSummaryMetricsSchema
    sources: list[AuditSummarySourceSchema]
    status: AuditSummaryStatusSchema


class DirectedModeStatsSchema(BaseModel):
    divergent_count: int
    document_count: int
    matching_count: int
    occurrence_count: int
    page_count: int


class DirectedModeEntrySchema(BaseModel):
    bbox: Any | None
    context: str
    document_id: int | None
    expected_value: str | None
    field_name: str | None
    filename: str
    kind: str
    page: int | None
    replacement_preview: str | None
    severity: str
    value: str


class DirectedModeOutputSchema(BaseModel):
    entries: list[DirectedModeEntrySchema]
    expected: str | None
    field_label: str | None
    mode: str
    query: str | None
    replace: str | None
    stats: DirectedModeStatsSchema
    summary: str
    title: str


class PackageMapSectionSchema(BaseModel):
    document_id: int
    document_filename: str
    end_page: int
    ld_codes: list[str]
    ld_page: int | None
    ld_row_count: int
    scope_id: int
    section_label: str
    section_type: str
    sheet_codes: list[str]
    sheet_count: int
    start_page: int
    title: str


class PackageMapDocumentSchema(BaseModel):
    classification: str
    discipline: str | None
    document_id: int
    filename: str
    page_count: int
    sections: list[PackageMapSectionSchema]
    tipo: str
    tomo: str | None
    volume: str | None


class PackageMapStatsSchema(BaseModel):
    document_count: int
    ld_section_count: int
    section_count: int
    sheet_count: int


class PackageMapSchema(BaseModel):
    documents: list[PackageMapDocumentSchema]
    identity: PackageSummaryIdentitySchema
    stats: PackageMapStatsSchema


class AiReviewContextSchema(BaseModel):
    document_id: int
    evidence_text: str
    filename: str
    kind: str
    page_end: int
    page_start: int
    section_label: str
    title: str


class AiReviewSuggestionSchema(BaseModel):
    category: str
    message: str
    reason: str
    severity: str
    source: str


class AiReviewStatsSchema(BaseModel):
    context_count: int
    needs_review_count: int
    probable_issue_count: int
    suggestion_count: int


class AiReviewSchema(BaseModel):
    contexts: list[AiReviewContextSchema]
    identity: PackageSummaryIdentitySchema
    mode: str
    provider_status: str
    summary: str
    suggestions: list[AiReviewSuggestionSchema]
    stats: AiReviewStatsSchema


class PageMapPageSchema(BaseModel):
    confidence: str
    discipline_code: str | None
    discipline_label: str | None
    discipline_type: str | None
    document_id: int
    evidence_text: str
    filename: str
    page: int
    page_type: str
    page_type_label: str
    scope_id: int | None
    signals: list[str]


class PageMapDocumentSchema(BaseModel):
    document_id: int
    filename: str
    page_count: int
    pages: list[PageMapPageSchema]
    tipo: str


class PageMapStatsSchema(BaseModel):
    document_count: int
    low_confidence_count: int
    page_count: int
    page_type_counts: dict[str, int]


class PageMapSchema(BaseModel):
    documents: list[PageMapDocumentSchema]
    stats: PageMapStatsSchema


class DrawingListRowSchema(BaseModel):
    description: str
    document_code: str
    item: str
    page: int
    source_text: str


class FooterAuditIdentitySchema(BaseModel):
    project_code: str | None
    work_name: str | None


class FooterAuditOccurrenceSchema(BaseModel):
    document_id: int
    field: str
    field_label: str
    filename: str
    normalized_value: str
    page: int
    source_text: str
    value: str


class FooterAuditFindingSchema(BaseModel):
    category: str
    field: str
    message: str
    occurrences: list[FooterAuditOccurrenceSchema]
    reason: str
    severity: str


class FooterAuditStatsSchema(BaseModel):
    document_count: int
    footer_page_count: int
    needs_review_count: int
    occurrence_count: int
    probable_issue_count: int


class FooterAuditSchema(BaseModel):
    findings: list[FooterAuditFindingSchema]
    identity: FooterAuditIdentitySchema
    occurrences: list[FooterAuditOccurrenceSchema]
    stats: FooterAuditStatsSchema


class DrawingListDocumentSchema(BaseModel):
    document_id: int
    filename: str
    project_codes: list[str]
    row_count: int
    rows: list[DrawingListRowSchema]
    tipo: str


class DrawingListStatsSchema(BaseModel):
    alert_count: int
    document_count: int
    row_count: int


class DrawingListAlertSchema(BaseModel):
    description: str
    document_code: str
    filename: str
    item: str
    message: str
    page: int
    severity: str
    source_text: str
    type: str


class DrawingListsSchema(BaseModel):
    alerts: list[DrawingListAlertSchema]
    lists: list[DrawingListDocumentSchema]
    stats: DrawingListStatsSchema


class DetectedSheetSchema(BaseModel):
    description: str | None
    item: str | None
    page: int
    sheet_code: str
    source_text: str


class DetectedSheetDocumentSchema(BaseModel):
    document_id: int
    filename: str
    sheet_count: int
    sheets: list[DetectedSheetSchema]
    tipo: str


class DetectedSheetsStatsSchema(BaseModel):
    document_count: int
    sheet_count: int


class DetectedSheetsSchema(BaseModel):
    documents: list[DetectedSheetDocumentSchema]
    stats: DetectedSheetsStatsSchema


class LdSheetMatchedSheetSchema(BaseModel):
    description: str | None
    filename: str
    item: str | None
    page: int
    scope_id: int | None = None
    sheet_code: str
    source_text: str


class LdMatchedRowSchema(BaseModel):
    description: str
    document_code: str
    filename: str
    item: str
    page: int
    scope_id: int | None = None
    source_text: str


class LdSheetCrosscheckResultSchema(BaseModel):
    category: str
    ld_description: str
    ld_document_code: str
    ld_filename: str
    ld_item: str
    ld_page: int
    ld_scope_id: int | None = None
    ld_source_text: str
    matched_sheet: LdSheetMatchedSheetSchema | None
    message: str
    reason: str
    severity: str
    type: str


class DetectedSheetCrosscheckResultSchema(BaseModel):
    category: str
    matched_ld_row: LdMatchedRowSchema | None
    message: str
    reason: str
    severity: str
    sheet_code: str
    sheet_description: str | None
    sheet_filename: str
    sheet_item: str | None
    sheet_page: int
    sheet_scope_id: int | None = None
    sheet_source_text: str
    type: str


class LdSheetCrosscheckStatsSchema(BaseModel):
    attention_count: int
    compatible_count: int
    combined_extraction_limit_count: int
    combined_needs_review_count: int
    combined_probable_issue_count: int
    extraction_limit_count: int
    needs_review_count: int
    ok_count: int
    probable_issue_count: int
    relevant_count: int
    reverse_extraction_limit_count: int
    reverse_needs_review_count: int
    reverse_other_document_count: int
    reverse_other_section_count: int
    reverse_probable_issue_count: int
    reverse_total_count: int
    total_count: int
    undeclared_sheet_count: int


class LdSheetCrosscheckSchema(BaseModel):
    results: list[LdSheetCrosscheckResultSchema]
    reverse_results: list[DetectedSheetCrosscheckResultSchema]
    stats: LdSheetCrosscheckStatsSchema


class MemorialAuditIdentitySchema(BaseModel):
    bairro: str | None
    municipality: str | None
    project_code: str | None
    work_name: str | None


class MemorialAuditOccurrenceSchema(BaseModel):
    document_id: int
    field: str
    field_label: str
    filename: str
    normalized_value: str
    page: int
    source_text: str
    value: str


class MemorialAuditFindingSchema(BaseModel):
    category: str
    field: str
    message: str
    occurrences: list[MemorialAuditOccurrenceSchema]
    reason: str
    severity: str


class MemorialAuditStatsSchema(BaseModel):
    document_count: int
    extraction_limit_count: int
    needs_review_count: int
    occurrence_count: int
    probable_issue_count: int


class MemorialAuditSchema(BaseModel):
    findings: list[MemorialAuditFindingSchema]
    identity: MemorialAuditIdentitySchema
    occurrences: list[MemorialAuditOccurrenceSchema]
    stats: MemorialAuditStatsSchema
